### PR TITLE
Report task outcome statistics after build finishes

### PR DESCRIPTION
**Pull request by eriwen**
#### Before
screen shot 2017-04-04 at 2 19 36 pm

#### After
screen shot 2017-04-04 at 3 39 45 pm

### Contributor Checklist
- [x] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [X] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [X] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [X] Ensure that tests pass locally: `./gradlew quickCheck <impacted-subproject>:check`

[CI Build](https://builds.gradle.org/viewQueued.html?itemId=2674208&tab=queuedBuildOverviewTab)

### Gradle Core Team Checklist
- [x] Verify design and implementation 
- [x] Verify test coverage and CI build status
- [x] Verify documentation including proper use of `@since` and `@Incubating` annotations for all public APIs

### CODE REVIEW

File: subprojects/core/src/main/java/org/gradle/api/internal/tasks/TaskStateInternal.java [hunk 1]
```diff
@@ -122,4 +123,12 @@ public boolean getNoSource() {
     public boolean isFromCache() {
         return outcome == TaskExecutionOutcome.FROM_CACHE;
     }
+
+    public boolean isActionsWereExecuted() {
```
Comment (w25r) [inline subprojects/core/src/main/java/org/gradle/api/internal/tasks/TaskStateInternal.java]: This reads pretty gnarly... I'm assuming we're keeping the 'is' to make the groovy syntax easier?

Could we have another method just for readability without the is?

Comment (eriwen) [inline subprojects/core/src/main/java/org/gradle/api/internal/tasks/TaskStateInternal.java]: The `is` prefix is intended to indicate the boolean return type, but I don't feel strongly about this and can change it if you do.

Comment (w25r) [inline subprojects/core/src/main/java/org/gradle/api/internal/tasks/TaskStateInternal.java]: I think it's fine as is... it just makes me read it very carefully.  And that's not all bad.

**Comment by lptr** [comment]
I think I like the changes overall. A couple things come up:

Are we okay with changing the long-standing output of `BUILD SUCCESSFUL`/`BUILD FAILED` that people might be grepping for? Could be a breaking change, but it's okay.

The thing I'm not fully comfortable is how we currently use the word `AVOIDED`:

```text
67 actionable tasks: 12 (18%) AVOIDED, 55 (82%) EXECUTED
```

I see a number of issues with this:

* We are introducing two new terms: `actionable` tasks and `AVOIDED` tasks. So reading this line can be hard for anyone seeing it for the first time. We should make every effort to make their job easier.
* We put numbers and percentages right next to each other with a monotype font, which is kinda hard to parse visually. It kind of looks like this when you're skimming through: `12 18 AVOIDED 55 82 EXECUTED`.
* I don't think declaring the percentage for both `AVOIDED` and `EXECUTED` provides any useful information. What we care about here is the amount of execution avoided.
* Likewise, having the sum number of actionable tasks reported, and then reporting both the number of executed, and the number of (sum - executed) as well seems redundant, too.

This would read a lot easier for me:

```text
67 actionable tasks: 55 EXECUTED, 12 AVOIDED (18%)
```

This would do away with different kind of numbers following each other directly, and would use the already known term `EXECUTED` before the new term `AVOIDED`.

Or even this:

```text
67 actionable tasks - 12 tasks avoided (18%)
```

Or maybe even skip the percentage:

```text
67 actionable tasks (12 tasks avoided)
```

I'm not fully happy with either, mostly because I don't completely like "avoided" as a term.

**Comment by eriwen** [comment]
@lptr Thanks very much for the analysis. We aren't omitting BUILD SUCCESSFUL/FAILED at all, so we should still pass the `grep` test. If they're looking for a newline afterward... well 4.0 is the season for breaking changes is my only response there.

I like your suggestion: `67 actionable tasks: 55 EXECUTED, 12 AVOIDED (18%)`

I could not come up with a better adjective than "actionable" for "non-skipped tasks with actions". Open to suggestions if anyone feels strongly here. Users should be pointed to the build scan for further information. 

On that note, I simply want to be consistent with our terminology in build scans regarding the word "avoided". Again, a user who wants more detail on what avoided means should look at their build scan, though I can see an argument for some kind of Glossary in the docs. I suppose a note in the release notes is appropriate as well.

**Comment by eriwen** [comment]
@lptr I also want to specifically request you to verify I removed what you expected from the cache stats reporting.

**Comment by lptr** [comment]
I'd personally go further with removing redundancy, but this is already a lot better.

What is our rationale for using all-caps for `EXECUTED` and `AVOIDED`? They look like `UP-TO-DATE` and `FROM-CACHE` labels we put next to the task paths themselves, but we never actually use either `EXECUTED` or `AVOIDED` as a task label. What do you think about:

```text
67 actionable tasks: 55 executed, 12 avoided (18%)
```

**Comment by eriwen** [comment]
@lptr Personally, I think it looks _cooler_ as all caps but that's not a strong justification. 

I want to note that we intend to change the default logging behavior in the next 2 releases such that tasks and their outcomes are no longer logged by default. Only _interesting_ ones are logged.

**Comment by eriwen** [comment]
@w25r please let me know when your review is complete

**Comment by w25r** [comment]
Looks like we have a problem if only one task is run.

```
BUILD SUCCESSFUL in 41s
1 actionable tasks: 1 executed, 0 avoided (0%)

```
Also, I still find 'avoided' a bit confusing.  What if they don't use build scans and have never seen that word before?

**Comment by w25r** [comment]
Minor note, I also think the actionable tasks line should precede the 'BUILD SUCCESSFUL' or 'BUILD FAILED' line.

**Comment by w25r** [comment]
One final note, although maybe simply a curiosity:  I would expect when re-running a task for the same number of actionable tasks, but a higher number avoided.  What I'm actually seeing is that usually there are MORE actionable tasks, though most or all avoided.

```
$ /gradle build

BUILD SUCCESSFUL in 0s
7 actionable tasks: 5 executed, 2 avoided (29%)
$ /gradle build

BUILD SUCCESSFUL in 0s
10 actionable tasks: 0 executed, 10 avoided (100%)

```

**Comment by eriwen** [comment]
@w25r I kicked off another build with some fixes: https://builds.gradle.org/viewQueued.html?itemId=2681419&tab=queuedBuildOverviewTab

Re: inconsistency. Build scans, please ;)

**Comment by w25r** [comment]
I'd still like to understand this discrepancy:

```gradle build
:compileJava
:processResources NO-SOURCE
:classes
:jar
:startScripts
:distTar
:distZip
:assemble
:compileTestJava NO-SOURCE
:processTestResources NO-SOURCE
:testClasses UP-TO-DATE
:test NO-SOURCE
:check UP-TO-DATE
:build

BUILD SUCCESSFUL in 0s
7 actionable tasks: 5 executed, 2 avoided (29%)
$ /tmp/gradle-pr/bin/gradle build
:compileJava UP-TO-DATE
:processResources NO-SOURCE
:classes UP-TO-DATE
:jar UP-TO-DATE
:startScripts UP-TO-DATE
:distTar UP-TO-DATE
:distZip UP-TO-DATE
:assemble UP-TO-DATE
:compileTestJava NO-SOURCE
:processTestResources NO-SOURCE
:testClasses UP-TO-DATE
:test NO-SOURCE
:check UP-TO-DATE
:build UP-TO-DATE

BUILD SUCCESSFUL in 0s
10 actionable tasks: 0 executed, 10 avoided (100%)
Gradles-MBP:gradle-build Gradle$ 
```

Is it about how tasks are counted if they are up-to-date and have no actions maybe?

**Comment by w25r** [comment]
Pending the CI run finishing and passing, I think it's good to go!

**Comment by eriwen** [comment]
Merged with 4b5be8babc15ad4d95b0033dda5c90fe2c1e90bd
